File: csharp/main/DecoratorDef.cs
using System;
using System.Collections.Generic;
using JetBrains.Annotations;

namespace Deltix.Luminary
{
	public sealed class DecoratorDef : CompositeTypeScope<DecoratorPropertyDef, TypeDecorator>, IConstantAndTypeContainer
	{
		public DecoratorDef([NotNull] ITypeScope parent, [NotNull] String name, [CanBeNull, ItemNotNull] List<String> comments = null)
			: base(name, parent, comments)
		{
			Type = new TypeDecorator(this);
			FullName = parent is FileDef ? Name : $"{((ItemDef) parent).FullName}.{Name}";

			parent.DefinedDecorators.Add(this);
		}

		public override TypeDecorator Type { get; }

		[NotNull, ItemNotNull]
		public List<ConstantDef> DefinedConstants { get; } = new List<ConstantDef>();

		public override String FullName { get; }
	}
}

File: java/src/main/java/deltix/luminary/DecoratorDef.java
package deltix.luminary;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;

public final class DecoratorDef extends CompositeTypeScope<DecoratorPropertyDef, TypeDecorator> implements ConstantAndTypeScope {
    private final TypeDecorator type;
    private final List<ConstantDef> definedConstants = new ArrayList<>();

    public DecoratorDef(@NotNull TypeScope parent, @NotNull String name, @Nullable List<String> comments) {
        super(parent, name, comments);
        this.type = new TypeDecorator(this);

        parent.getDefinedDecorators().add(this);
    }

    public DecoratorDef(@NotNull TypeScope parent, @NotNull String name) {
        this(parent, name, null);
    }

    @Override
    @NotNull
    public TypeDecorator getType() {
        return type;
    }

    @Override
    public List<ConstantDef> getDefinedConstants() {
        return definedConstants;
    }
}
